### PR TITLE
David Loveday

### PR DESCRIPTION
**Pull request by AquilaVirtual**

### CODE REVIEW

**Comment by mixelpixel** [comment]
@AquilaVirtual 👋 

It's nice to see that you are committing frequently. You get a gold ⭐️ !!

Your git log _**tells the story of your development adventure**_ but I'd ❤️ to see more 😀 

While the Git commit titles are short, you can add a "body" to the messages which can be much more detailed. BEFORE you add the closing quote to the commit message, hit enter twice to add a new line between the title and the body. For example:

```
$ git commit -m "This is a Git commit TITLE less than 72 characters

This is a Git commit BODY which can contain as much pertinent information and descriptions as you see fit. Some development teams like tons and tons of information, and depending on the scope of the commit, it may be necessary to make lots of detailed notes - especially in projects like React where there can be a vast amount of files and a complex project architecture.
- I did this.
- I did that.
- this happens.
- that needs work.
- this thing got refactored.
- where's my Lambda swag?????"
```

...on GitHub in the Pull Requests, the body will be accessible by clicking on the three little dots:
screen shot 2018-04-12 at 5 32 06 am

...such that they will reveal the BODY:
screen shot 2018-04-12 at 5 32 20 am

If you're not already in the #git channel in Slack, come check it out: https://lambdaschoolstudents.slack.com/archives/C5Z8BMX6K/p1523565197000063

😺

File: ANSWERS.md [hunk 1]
```diff
@@ -1,6 +1,16 @@
 # Your responses to the short answer questions should be laid out here using Mark Down.
 1. Describe the biggest difference between `.forEach` & `.map`.
+   `.ForEach` does not return anything, it returns undefined and `.map` returns a new array.
```
Comment (JasonCorp84) [inline ANSWERS.md]: 👍

File: ANSWERS.md [hunk 2]
```diff
@@ -1,6 +1,16 @@
...
 2. What is the difference between a function and a method?
+A function is its own entity while a method belongs to an object
 3. What is closure?
+Closure is when a function or an object has access to somthing outside its scope.
```
Comment (JasonCorp84) [inline ANSWERS.md]: 👍

File: ANSWERS.md [hunk 3]
```diff
@@ -1,6 +1,16 @@
...
 2. What is the difference between a function and a method?
+A function is its own entity while a method belongs to an object
 3. What is closure?
+Closure is when a function or an object has access to somthing outside its scope. 
 4. Describe the four rules of the 'this' keyword.
+  1. If a function is called as a method, `this` is the object that the function is a property of.
+  2. `this` is global object when a called with a free function
+  3. If the `new` keyword is used when calling a function, `this` inside the function is a brand new object.
+  4. If `apply`, `call`, or `bind` are used to call a function, `this` inside the function is the object that is passed in as the argument.
+
```
Comment (JasonCorp84) [inline ANSWERS.md]: 👍 👍 👍 👍

File: challenges/functions.js [hunk 4]
```diff
@@ -7,27 +7,42 @@
...
+  return param1 + param2;
+}
 
+function multiply (param1, param2) {
+  return param1 * param2;
+}
 
+function greeting(param1, param2) {
+  return `Hello ${param1} ${param2}, nice to meet you!`;
+}
```
Comment (JasonCorp84) [inline challenges/functions.js]: 👍 👍 👍

File: challenges/objects-arrays.js [hunk 5]
```diff
@@ -13,6 +13,29 @@
...
+let tyrannosaurus = {
+  "diet": "carnivorous", 
+  "weight": "7000kg",
+  "length": "12m",
+  "period": "Late Cretaceious",
+  "roar"  : function() {
+    return "RAWERSRARARWERSARARARRRR!";
+  }
+}
+
```
Comment (JasonCorp84) [inline challenges/objects-arrays.js]: 👍

File: challenges/objects-arrays.js [hunk 6]
```diff
@@ -13,6 +13,29 @@
...
+  "length": "9m",
+  "period": "Late Jurassic"
+}
+
+let velociraptor = {
+  "diet": "carnivorous",
+  "weight": "15kg",
+  "length": "1.8m",
+  "period": "Late Cretaceious"
+}
```
Comment (JasonCorp84) [inline challenges/objects-arrays.js]: 👍

File: challenges/objects-arrays.js [hunk 7]
```diff
@@ -46,6 +69,9 @@ const graduates = [{"id":1,"first_name":"Cynde","university":"Missouri Southern
 
 Once you have the new array created, sort the universities alphabetically and log the result. */
 const universities = [];
+for(let i = 0; i < graduates.length; i++) {
+  universities.push(graduates[i].university);
+}
 console.log(universities)
```
Comment (JasonCorp84) [inline challenges/objects-arrays.js]: let's not forget to sort the array by alphabet

File: challenges/objects-arrays.js [hunk 8]
```diff
@@ -55,11 +81,21 @@ Name email@example.com
 
 Log the result of your new array. */
 const contactInfo = [];
+for(let i = 0; i < graduates.length; i++) {
+  contactInfo.push(`${graduates[i].first_name} ${graduates[i].email}`);
+}
 console.log(contactInfo);
```
Comment (JasonCorp84) [inline challenges/objects-arrays.js]: 👍

File: challenges/objects-arrays.js [hunk 9]
```diff
@@ -55,11 +81,21 @@ Name email@example.com
...
 /* Request 3: Find out how many universities have the string "Uni" included in their name. Create a new array called uni that contains them all. Log the result. */
 const uni = [];
+
+for(let i = 0; i < graduates.length; i++) {
+  if(graduates[i].university ===/Uni(,)\1/) {
+  uni.push(graduates[i].university);
+  }
+}
+
 console.log(uni);
```
Comment (JasonCorp84) [inline challenges/objects-arrays.js]: Ok, but how many?

File: challenges/objects-arrays.js [hunk 10]
```diff
@@ -80,8 +116,14 @@ zooAnimals = [{"animal_name":"Jackal, asiatic","population":5,"scientific_name":
 
 // The zoos need a list of all their animal's names converted to lower case.  Create a new array named lowerCase and map over each name to convert them all to lower case.  Log the resut.
 let lowerCase = [];
+for(let i = 0; i < zooAnimals.length; i++) {
+  lowerCase.push(zooAnimals[i].animal_name.toLowerCase());
+}
```
Comment (JasonCorp84) [inline challenges/objects-arrays.js]: 👍

File: challenges/objects-arrays.js [hunk 11]
```diff
@@ -80,8 +116,14 @@ zooAnimals = [{"animal_name":"Jackal, asiatic","population":5,"scientific_name":
...
 console.log(lowerCase); 
 
 // The zoos need to know their total animal population across the United States.  Add up all the population numbers from all the zoos using the .reduce() method.
 let populationTotal = [];
+populationTotal = zooAnimals.reduce(function(acc, val){
+return acc + val.population;
+}, 0);
 console.log(populationTotal);
```
Comment (JasonCorp84) [inline challenges/objects-arrays.js]: 👍

File: challenges/prototypes.js [hunk 12]
```diff
@@ -6,28 +6,48 @@
   Create a constructor function named CuboidMaker that accepts properties for length, width, and height
 */
 
+function CuboidMaker (properties) {
+  this.length = properties.length;
+  this.width = properties.width;
+  this.height = properties.height;
+}
```
Comment (JasonCorp84) [inline challenges/prototypes.js]: 👍

File: challenges/prototypes.js [hunk 13]
```diff
@@ -6,28 +6,48 @@
...
 
 /* == Step 2: Volume Method ==
   Create a method using CuboidMaker's prototype that returns the volume of a given cuboid's length, width, and height
   
   Formula for cuboid volume: length * width * height
 */
 
+CuboidMaker.prototype.volume = function() {
+  return this.length * this.width * this.height; 
+}
```
Comment (JasonCorp84) [inline challenges/prototypes.js]: 👍

File: challenges/prototypes.js [hunk 14]
```diff
@@ -6,28 +6,48 @@
...
 
 /* == Step 3: Surface Area Method ==
   Create another method using CuboidMaker's prototype that returns the surface area of a given cuboid's length, width, and height. 
 
   Formula for cuboid surface area of a cube: 2 * (length * width + length * height + width * height)
 */
 
+CuboidMaker.prototype.surfaceArea = function() {
+  return  2 * (this.length * this.width + this.length * this.height + this.width * this.height); 
+}
```
Comment (JasonCorp84) [inline challenges/prototypes.js]: 👍

File: challenges/prototypes.js [hunk 15]
```diff
@@ -6,28 +6,48 @@
...
 */
 
+function Cuboid (properties) {
+  CuboidMaker.call(this, properties);
+}
+
+Cuboid.prototype = Object.create(CuboidMaker.prototype);
+
+let cuboid = new Cuboid({"length": 4, "width": 5, "height": 5});
+
```
Comment (JasonCorp84) [inline challenges/prototypes.js]: 👍

File: challenges/classes.js [hunk 16]
```diff
@@ -8,4 +8,48 @@
...
+  this.height = properties.height;
+  }
+  volume() {
+    return this.length * this.width * this.height; 
+  };
+
+  surfaceArea() {
+    return  2 * (this.length * this.width + this.length * this.height + this.width * this.height); 
+  };
+}
```
Comment (JasonCorp84) [inline challenges/classes.js]: 👍

File: challenges/classes.js [hunk 17]
```diff
@@ -54,3 +54,22 @@ let cuboid = new Cuboid({"length": 4, "width": 5, "height": 5});
...
+  };
+
+  volume() {
+    return  (this.side * this.side * this.side);
+  };
+}
+
+
+let cube = new CubeMaker({"side": 3});
+
```
Comment (JasonCorp84) [inline challenges/classes.js]: 👍

**Comment by JasonCorp84** [comment]
David,

### Self Study Questions
[x]  Answered all 

### Challenges
[ x] Objects 
  [ x] Create 3 dinosaur objects
  [ x] Test your objects using console logs
  [ x] Create rawr method

[ x] Arrays
  [ x] Part 1 - Complete all requests without using methods
  [ x] Part 2 - Complete all requests using methods

[ x] Functions
  [ x] Callbacks
  [ x] Closure

[ x] Prototypes
  [ x] Base Constructor
  [ x] Volume Method
  [ x] Surface Area Method
  [ x] Create new object that uses CuboidMaker

[ x] Classes
  [ x] Bring in prototype
  [ x] Test using console logs
  [ x] Extend base class with sub class

## Good Habits
  [ x] Pull Request title is your first and last name (add any notes in the comment field)
  [ x] Code has been committed regularly
  [ x] Commit comments are informative
  [ x] Code is organized in a logical way

You have shown great progress in the last 2 weeks.It is easy to read your code.  Excellent job! Keep it up!

{3}

File: challenges/objects-arrays.js (hunk 9 above)
Comment (AquilaVirtual) [reply challenges/objects-arrays.js]: Instruction didn't say to console.log how many, but I just did.

**Pull request closed without merging**